src/test/clojure/xtdb/operator/scan_test.clj

```diff
@@ -676,3 +679,56 @@
                             [:put-docs :docs {:xt/id id}]))
 
   (t/is (= 2000 (count (xt/q tu/*node* '(from :docs [{:xt/id id}]))))))
+
+(deftest test-recency-filtering
+  (tu/with-tmp-dirs #{node-dir}
+    (let [[instant-src clock] (tu/->mock-clock-pair (tu/->instants :year))]
+      (binding [expr/*clock* clock
+                c/*page-size* 1
+                c/*l1-file-size-rows* 256
+                c/*ignore-signal-block?* true]
+        (with-open [node (xtn/start-node {:log [:local {:path (.resolve node-dir "log"), :instant-src instant-src}]
+                                          #_[:in-memory {:instant-src (tu/->mock-clock (tu/->instants ChronoUnit/YEARS))}]
+                                          :storage [:local {:path (io/file (.toFile node-dir) "objects")}]
+                                          :indexer {:rows-per-chunk 1}})]
+          ;; 2020 - 2024
+          (doseq [i (range 5)]
+            (xt/execute-tx node [[:put-docs :docs {:xt/id 1 :col i}]]))
+
+          (tu/finish-chunk! node)
+          ;; compaction happens in 2025
+          (c/compact-all! node)
+
+          (let [->old-merge-tasks scan/->merge-tasks
+                merge-tsk-cnt (atom 0)]
+            (with-redefs [scan/->merge-tasks (fn [& args]
+                                               (let [res (apply ->old-merge-tasks args)]
+                                                 (reset! merge-tsk-cnt (-> res first :leaves count))
+                                                 res))]
+
+              ;; no filter,
+              (t/is (= 1 (count (tu/query-ra '[:scan {:table docs} [xt$id]] {:node node}))))
+              ;; 2024 + inf
+              (t/is (= 2 @merge-tsk-cnt))
+
+              ;; two entries 2020 and 2021
+              (t/is (= 2 (count (tu/query-ra '[:scan {:table docs :for-valid-time [:between #inst "2023" #inst "2025"]}
+                                               [xt$id xt$valid_from xt$valid_to xt$system_from xt$system_to]]
+                                             {:node node}))))
+
+              ;; 2023, 2024 and inf
+              (t/is (= 3 @merge-tsk-cnt))
+
+
+              ;; newest entry, basis at 2024
+              (t/is (= 1
+                       (count (tu/query-ra '[:scan {:table docs :for-valid-time [:between #inst "2025" nil]} [xt$id]]
+                                           {:node node}))))
+              ;; 2024 and inf
+              (t/is (= 2 @merge-tsk-cnt))
+
+
+              ;; everything
+              (t/is (= 5
+                       (count (tu/query-ra '[:scan {:table docs :for-valid-time :all-time} [xt$id]] {:node node}))
+                       @merge-tsk-cnt)))))))))
```

Comment:
This test is a bit of a mouthful (or however you would say that for a test). So maybe we can find something simpler.

Reply:
as discussed, let's remove the with-redefs here, just test the final answer, and rely on the unit tests below to confirm that it got there in the correct manner.

value of these tests is in showing that, even for cases which might cause the algo trouble, we still return the correct answers - i.e. we haven't missed a page